# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1296120173 18000
#      Thu Jan 27 04:22:53 2011 -0500
# Branch release-3-4-x
# Node ID 077ece977a7b6a15d3207fe67912c89b9b214ce8
# Parent  4f8069727e9f788c0e781b640f30fb6d8c8553b3
fix construction of class exemplars

diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,15 @@
+2011-01-27  John W. Eaton  <jwe@octave.org>
+
+	* ov-class.cc (octave_class::reconstruct_exemplar):
+	Call constructor with do_multi_index_op instead of feval.
+	Verify that symbol_table::find_method did return a class
+	constructor for the correct type.
+
 2011-01-26  John W. Eaton  <jwe@octave.org>
 
 	* lex.ll, graphics.cc, gl-render.cc, ov-mex-fcn.h, graphics.h.in:
 	Delete trailing whitespace.
 
 2011-01-26  John W. Eaton  <jwe@octave.org>
 
 	Bug #32242.
diff --git a/src/ov-class.cc b/src/ov-class.cc
--- a/src/ov-class.cc
+++ b/src/ov-class.cc
@@ -1063,19 +1063,35 @@ octave_class::reconstruct_exemplar (void
     = octave_class::exemplar_map.find (c_name);
 
   if (it != octave_class::exemplar_map.end ())
     retval = true;
   else
     {
       octave_value ctor = symbol_table::find_method (c_name, c_name);
 
-      if (ctor.is_defined ())
+      bool have_ctor = false;
+
+      if (ctor.is_defined () && ctor.is_function ())
         {
-          octave_value_list result = feval (ctor, 1);
+          octave_function *fcn = ctor.function_value ();
+
+          if (fcn && fcn->is_class_constructor (c_name))
+            have_ctor = true;
+
+          // Something has gone terribly wrong if
+          // symbol_table::find_method (c_name, c_name) does not return
+          // a class constructor for the class c_name...
+          assert (have_ctor);
+        }
+
+      if (have_ctor)
+        {
+          octave_value_list result
+            = ctor.do_multi_index_op (1, octave_value_list ());
 
           if (result.length () == 1)
             retval = true;
           else
             warning ("call to constructor for class %s failed", c_name.c_str ());
         }
       else
         warning ("no constructor for class %s", c_name.c_str ());
